flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -195,6 +195,8 @@ public Dispatcher(
         this.dispatcherBootstrapFactory = checkNotNull(dispatcherBootstrapFactory);
 
         this.recoveredJobs = new HashSet<>(recoveredJobs);
+        this.blobServer.retainJobs(
```

Comment:
should this be done in the io executor?

Reply:
Yes, I think this is required. Good catch!

Reply:
Hmm, thinking again about it I think that the `ioExecutor` won't work. The problem is that we want to finish the cleanup before the `Dispatcher` starts. Otherwise, we risk that we remove jobs that were newly submitted. Hence, I will revert this change. I also think that it is not a problem to not run it in the `ioExecutor` because the `Dispatcher` is not running at this point in time.

Reply:
What you could do is first determine the set of jobs to be cleaned up synchronously, and then issue the cleanup in the `ioExecutor`.

Reply:
But I think it should be fine either way.